stack.md

```diff
@@ -0,0 +1,64 @@
+# Yak Stack for Rovers
+
+Yak Rovers act as a collective. Rovers, humans and non-roving applications interact using Web APIs.
+
+## Connectivity
+
+It's hard to predict the communication infrastructure that could be available
+for Mars missions. To enable building for today we priorize common simple
+mechanisms for communicating between Rovers and applications.
+
+Each Rover is expected to have an IP address. Services are offered over HTTP.
+
+Networks will be presumed secure. This is accomplished either physically (such
+as a 140 million mile airgap to Mars) or via VPNs routed over the public
+internet.
+
+## VPN
+
+On Earth, Rovers communicate over secured VPN powered by
+[Tailscale](https://tailscale.com) (0-configuration WireGuard) which is offered
+for free for personal use as well as for Open Source GitHub communities.
+
+## API
+
+The HTTP-based API for interacting with Rovers should be expected to evolve over time. 
+
+Most interactions should be done with `Content-Type` of `json`.
+
+To start Rovers should provide at least:
+
+### GET /v1
+
+Top level resource should respond with identifying information for the rover as
+well as links to resources.
+
+### POST /v1/CI
```

Comment:
One comment that came up in discussion in yesterday's meeting was about whether *all* communication should happen through POST to a single resource. The idea being to replace REST semantics with RPC semantics.

This is certainly possible and comes with tradeoffs. As an example, I've gotten a lot of use out of [Twirp](https://github.com/twitchtv/twirp), but [gRPC](https://grpc.io) is probably more commonly considered. Both of those are about invoking RPC over HTTP. 

The suggestion was, rather than retrieve the image from a camera like:

    $ curl http://rover/v1/CAM1 -o cam1.jpg

We would invoke an RPC:

    $ curl  --request "POST" \
        --header "Content-Type: application/json" \
        --data '{"cam": "1"}' \
        http:/rover/v1/camera.SnapPhoto
       -o cam1.jpg

Certainly both are functional. I've found RPC mechanics to be preferable when high-efficiency is needed (twirp and gRPC are both based on protocol buffers) and the API needs to be very fixed and standardized. Typically an API provider will ship a language specific library for interacting with the API because, as you can see above, the interaction is slightly less friendly at the HTTP level.

REST-oriented protocols leverage (or perhaps overload, depending on your perspective) the natural semantics of HTTP to define an API. Baked into the protocol (and numerous client libraries) are verbs (PUT, GET, DELETE, POST) and resources (URLs).  A standard set of errors are provided. If I want to tell someone else how to retrieve a camera image, I simply provide a url rather than a URL and an API specification.

I think we should be optimizing for flexibility, ease-of-use and iteration speed. To me a REST-oriented API fits that bill, but I'm of course open to debate on this.

Reply:
I think the comment yesterday was more of a hybrid between REST and RPC-style interactions. Perhaps one of the reason from my side is that I have never seen a "properly" implemented fully-fledged REST API.

Yesterday was not really about RPC, because the single HTTP endpoint for all communications is seen as just a mail box without guarantee. The HTTP response must be 200 on completion, but the underlying action performed by the rover is up to the rover implementer. In fact, the 200 response could contain a payload for the format of the possible action answer---something related to REST. Something like this:

```mermaid
sequenceDiagram
autonumber
participant C as Client
participant S as Server
participant K as Store
participant Cam as Camera1
C->>+S: `/v1/ci/cam/1`
Note right of S: Endpoint discovered from `/v1` as REST
S-->>C: `{"status": 200, "pending_format": {...},"success_format": {...}, "error_format": {...}, "id": 123}`
S->>+Cam: Ask for frame
S->>+K: Request 123: Status waiting
K-->>-S: Noted
loop Poll
  C->>+S: `/v1/ci/123`
  S->>+K: Request 123: Status?
  K-->>-S: Pending
  S-->>-C: `{"status": 200, "type": "pending", "formatted_payload": {...}, "id": 123}`
end
Cam-->>-S: Send frame
S->>+K: Request 123: Status done, with frame
K-->>-S: Noted
C->>+S: `/v1/ci/123`
S->>+K: Request 123: Status?
K-->>-S: Done
S-->>-C: `{"status": 200, "type": "success", "formatted_payload": {...}, "id": 123}`
```

The poll loop is a waste, and a request could include a callback URL to the client.

Anyway the only motivation in the suggestion is, by my understanding, about requiring as small an API as possible. In retrospective, it just moves the effort elsewhere, so not convinced about the benefit at this time. A _proper_ REST framework fits the bill (the `GET /v1` is so often missing or incomplete :-D )!

Reply:
Ahh yes. I think we will have to pay special attention to establishing patterns for long running tasks. 

I think the expectation should be that responses are fast. A longer running task needs some other mechanism to communicate the result.

I particularly like the [blocking query pattern established in the consul api](https://www.consul.io/api-docs/features/blocking) where a client can opt in to have a request block until some change has occurred, but with no guarantees about the result. It's an elegant workaround for polling against some resource.